### ISSUE TITLE
RuntimeError: dictionary changed size during iteration in TaskManager.(timeout|cancelled)_tasks

### ISSUE THREAD

**Issue by charettes** [noxdafox/pebble#96]
The full traceback is

```
RuntimeError: dictionary changed size during iteration
  File "threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "pebble/pool/process.py", line 167, in pool_manager_loop
    pool_manager.update_status()
  File "pebble/pool/process.py", line 220, in update_status
    self.update_tasks()
  File "pebble/pool/process.py", line 225, in update_tasks
    for task in self.task_manager.timeout_tasks():
  File "pebble/pool/process.py", line 305, in timeout_tasks
    return tuple(t for t in tuple(self.tasks.values()) if self.timeout(t))
```

If `tuple(self.tasks.values())` happens to be reached while `self.tasks` is altered (e.g. via `register`, `task_start`, `task_done`) then it seems like it might result in this exception.

I'm not sure which conditions caused this but we noticed some `ProcessExpired('Abnormal termination')` logs around the time it happened which makes me believe there might have been concurrent `task_done` calls.

If `TaskManager` is meant to be thread-safe it should use a lock around interactions with `self.tasks` to prevent this from happening I guess? I was under the impression that `tuple(self.tasks.values())` would have been enough but it seems under some circumstances it's not?

**Comment by noxdafox** [comment]
Hello,

what version of Pebble are you using? The error traceback does not reflect the last versions.

**Comment by charettes** [comment]
This particular traceback is from 4.5.3 but the same access patterns are still in the `master` branch

https://github.com/noxdafox/pebble/blob/8fd612e5086cee71d61194ac9b33b46b5bba60dd/pebble/pool/process.py#L283-L318

**Comment by noxdafox** [comment]
Could you provide an example which reproduces the issue? I never observed such a problem.

**Comment by noxdafox** [comment]
Closing this issue due to lack of activity. Please re-open if you require further clarifications.

**Comment by PLPeeters** [comment]
@noxdafox Getting the same thing pretty randomly, both on 4.6 and 5.0.

It's weird because creating the `tuple` should be enough to avoid this, yet it still happens. I'm not too sure, but ChatGPT says this:

> The Python interpreter might be optimizing or handling these operations in a way that doesn't fully isolate the snapshot creation from the iteration, especially under the hood in CPython's implementation, where the error occurs.

Might be due to me using Pyston as well. The initial thing ChatGPT suggested was moving the snapshotting outside of the comprehension; maybe that might help it not get incorrectly optimised (if that is even the issue). Introducing a lock would be the other option.

**Comment by noxdafox** [comment]
`tuple(dict.values())` should be thread safe but on CPython. I am not sure how other implementations work in this matters.

I will include a safeguard.

**Comment by noxdafox** [comment]
Issue fixed in release `5.0.7`, please re-open this issue if the problem persists.
